systemd_ctypes/bus.py

```diff
@@ -320,124 +327,467 @@ def add_match(self, rule, handler):
+
+            a = MyObject()
+            a.counter = 5
+            a.counter += 1
+            print(a.counter)
+
+        The other way to define properties is with a getter function.  In this
+        case, you can read from the property in the normal way, but not write
+        to it.  You are responsible for emitting change signals for yourself.
+        You must not provide the value= kwarg.
+
+            class MyObject(bus.Object):
+                _counter = 0
+
+                counter = bus.Interface.Property('i')
+                @counter.getter
+                def get_counter(self):
+                    return self._counter
+
+                def set_counter(self, value):
+                    self._counter = value
+                    self.property_changed('Counter')
+
+        In either case, you can provide a setter function.  This function has
+        no impact on Python code, but makes the property writable from the view
+        of D-Bus.  Your setter will be called when a Properties.Set() call is
+        made, and no other action will be performed.  A trivial implementation
+        might look like:
+
+            class MyObject(bus.Object):
+                counter = bus.Interface.Property('i', value=0)
+                @counter.setter
+                def set_counter(self, value):
+                    # we got a request to set the counter from D-Bus
+                    self.counter = value
+
+        In all cases, the first (and only mandatory) argument to the
+        constructor is the D-Bus type of the property.
+
+        Your getter and setter functions can be provided by kwarg to the
+        constructor.  You can also give a name= kwarg to override the default
+        name conversion scheme.
+        """
+        _category = 'properties'
+
+        _getter: Optional[Callable[[Any], Any]]
+        _setter: Optional[Callable[[Any, Any], None]]
+        _type_string: str
+        _value: Any
+
+        def __init__(self, type_string: str,
+                     value: Any = None,
+                     name: Optional[str] = None,
+                     getter: Optional[Callable[[Any], Any]] = None,
+                     setter: Optional[Callable[[Any, Any], None]] = None):
+            assert value is None or getter is None, 'A property cannot have both a value and a getter'
+
+            super().__init__(name=name)
+            self._getter = getter
+            self._setter = setter
+            self._type_string = type_string
+            self._value = value
+
+        def _describe(self) -> Dict[str, Any]:
+            return {'type': self._type_string, 'flags': 'r' if self._setter is None else 'w'}
+
+        def __get__(self, obj: object, cls: Optional[type] = None) -> Any:
+            if obj is None:
+                return self
+            if self._getter is not None:
+                return self._getter.__get__(obj, cls)()
+            elif self._value is not None:
```

Comment:
Just for myself: `None` can never be a valid value for a D-Bus property of any type, right?

Reply:
Correct.  If we had managed to get Maybe types added back in the day, it would be a different story, indeed.